bin/codemods/src/assign-component-to-displayname-variable.js

```diff
@@ -50,91 +39,73 @@ const displayNamePropertyIdentifier = {
-
-	const exportDefaultDeclarations = root.find( j.ExportDefaultDeclaration );
-
-	exportDefaultDeclarations.forEach( node => {
-		const calleeNameValue = node.get( 'declaration', 'callee', 'name' ).value;
+	const defaultExportDeclaration = _.head( root.find( j.ExportDefaultDeclaration ).nodes() );
+	const hocIdentifier = _.get( defaultExportDeclaration, [ 'declaration', 'callee', 'name' ] );
 
-		if ( ! calleeNameValue ) {
-			// return early, no immediate function call
-			return;
-		}
+	const hocArgs = _.get( defaultExportDeclaration, [ 'declaration', 'arguments' ] );
+	const component = getComponentFromArgs( hocArgs );
+	const displayName = extractDisplayName( component, j );
 
-		const exportDefaultInstance = j( node );
+	// noop if the file does not have a default export of a react component that has a displayName
+	if ( ! defaultExportDeclaration || ! component || ! displayName ) {
+		return file;
+	}
 
-		const args = node.get( 'declaration', 'arguments' );
-		const matchingArg = getMatchingArg( args );
+	const isCreateClass = isCreateClassComponent( component );
 
-		if ( ! matchingArg ) {
-			// return early, no createClass or extends Component argument found
-			return;
-		}
-
-		const isCreateClassInstance = argIsCreateClassInstance( matchingArg );
-
-		const propertyType = isCreateClassInstance ? j.Property : j.ClassProperty;
-
-		const classExpressions = isCreateClassInstance
-			? exportDefaultInstance.find( j.CallExpression, createClassIdentifier )
-			: exportDefaultInstance.find( j.ClassExpression );
-
-		const displayNameValue = classExpressions
-			.find( propertyType, displayNamePropertyIdentifier )
-			.at( 0 )
-			.get( 'value', 'value' ).value;
-
-		if ( ! displayNameValue ) {
-			// return early, no displayNameValue
-			return;
-		}
-
-		exportDefaultInstance.replaceWith( node => [
-			isCreateClassInstance
+	return root
+		.find( j.ExportDefaultDeclaration )
+		.replaceWith( () => [
+			isCreateClass
 				? // if createClass, replace with const x = React.createClass...
 					j.variableDeclaration( 'const', [
-						j.variableDeclarator( j.identifier( displayNameValue ), matchingArg.value ),
+						j.variableDeclarator( j.identifier( displayName ), component ),
 					] )
 				: // else Replace with class x extends {original-super-class}...
-					j.classDeclaration(
-						j.identifier( displayNameValue ),
-						matchingArg.value.body,
-						matchingArg.value.superClass
-					),
-			// In both cases export with export default calleeName( x ),
+					j.classDeclaration( j.identifier( displayName ), component.body, component.superClass ),
+			// In both cases export with export default hoc( ...args, displayName, ...args ),
 			// maintaining argument position
 			j.exportDefaultDeclaration(
 				j.callExpression(
-					j.identifier( calleeNameValue ),
-					args.map( replaceClassOrGetValue( displayNameValue ) )
+					j.identifier( hocIdentifier ),
+					hocArgs.map( classToIdentifier( displayName, j ) )
```

Comment:
I like the clarity `hocArgs` brings :)

I wonder if there's a nicer way to handle functions that need `j` so that we can keep those functions unary.
Would it be reasonable to do something like this pattern?:
```js
const createSomeFunction = j => someArg => { ... }

export default function transformer( file, api ) {
  const someFunction = createSomeFunction( j )
  const someOtherFunction = createSomeOtherFunction( j )
```

More just a thought & question - I'm happy with what we have here.

Reply:
The reason I needed to add `j` as a param was because I wanted to hoist the helper function above `transformer`.  

I actually really like how you curried `replaceClassOrGetValue` so that it became unary and could be passed directly to `map`.  I tried to do keep the same style by making the return of `classToIdentifier` a unary function.  I'll play with moving it back into `transformer` to see if its clearer

Reply:
In the format of the pattern above we might end up with something like this:

```js
const createClassToIdentifier = j => displayName => arg => { ... }

export default function transformer( file, api ) {
  ...
  const classToIdentifier = createClassToIdentifier( j )

  ...

  hocArgs.map( classToIdentifier( displayName ) )

```

Perhaps that would be a bit 'curry heavy' though.
With this pattern we could hoist our functions, sort of 'bind' them (similar to calling `x.bind( this )` in a constructor) and then use them without caring for injecting dependencies like `j`.

Reply:
https://github.com/Automattic/wp-calypso/blob/add/codemod/assign-default-refactor--extra-bowl-of-curry/bin/codemods/src/assign-component-to-displayname-variable.js

I put together a branch with this on, feel free to ignore it (it really shouldn't be a blocker since it's more just some thinking-out-loud-ness) :)

Reply:
Making heavy use of currying here makes meaningful names very difficult to come up with. How do we name a function like `classToIdentifier`, a 3 arg function, at each stage of currying?  `createCreateClassToIdentifier` --> `createClassToIdentifier` --> `classToIdentifier`?

I'd personally be even more amenable to using the signatures you've made ( `j -> displayName -> arg` ) and then calling it like `func( j )( displayName )` but it all feels unwieldy to me